File: csharp/ql/test/query-tests/Security Features/CWE-730/RegexInjection/RegexInjection.cs
using System;
using System.Web;
using System.Text.RegularExpressions;

public class RegexInjectionHandler : IHttpHandler
{

    public void ProcessRequest(HttpContext ctx)
    {
        string regex = ctx.Request.QueryString["regex"]; // $ Source=r1
        string userInput = ctx.Request.QueryString["userInput"];

        // BAD: User input used as regex
        new Regex(regex).Match(userInput); // $ Alert=r1
        // GOOD: User input escaped before being used as regex
        new Regex(Regex.Escape(regex)).Match(userInput);
    }

    public bool IsReusable
    {
        get
        {
            return true;
        }
    }

}

File: java/ql/src/Security/CWE/CWE-400/RegexInjection.java
public boolean isUsernameInPassword(HttpServletRequest request) {
  String username = request.getParameter("username");
  String pasword = request.getParameter("password");

  // BAD: User provided input is being used to create regex without escaping
  Pattern p = Pattern.compile(username);
  Matcher m = p.matcher(password);
  return = m.matches();

  // GOOD: User provided input escaped before being used to create regex
  String safeUsername = Pattern.quote(username);
  Pattern p = Pattern.compile(safeUsername);
  Matcher m = p.matcher(password);
  return = m.matches();
}